src/common/ceph_crypto.cc

```diff
@@ -12,83 +12,122 @@
-static uint32_t crypto_refs = 0;
-static NSSInitContext *crypto_context = NULL;
-static pid_t crypto_init_pid = 0;
-
-void ceph::crypto::init(CephContext *cct)
+static void
+ssl_locking_callback(
+  const int mode,
+  const int mutex_num,
+  [[maybe_unused]] const char *file,
+  [[maybe_unused]] const int line)
 {
-  pid_t pid = getpid();
-  pthread_mutex_lock(&crypto_init_mutex);
-  if (crypto_init_pid != pid) {
-    if (crypto_init_pid > 0) {
-      SECMOD_RestartModules(PR_FALSE);
-    }
-    crypto_init_pid = pid;
+  if (mutex_num < 0 || static_cast<size_t>(mutex_num) >= ssl_mutexes.size()) {
+    ceph_assert_always("openssl passed wrong mutex index" == nullptr);
   }
 
-  if (++crypto_refs == 1) {
-    NSSInitParameters init_params;
-    memset(&init_params, 0, sizeof(init_params));
-    init_params.length = sizeof(init_params);
-
-    uint32_t flags = (NSS_INIT_READONLY | NSS_INIT_PK11RELOAD);
-    if (cct->_conf->nss_db_path.empty()) {
-      flags |= (NSS_INIT_NOCERTDB | NSS_INIT_NOMODDB);
+  if (mode & CRYPTO_READ) {
+    if (mode & CRYPTO_LOCK) {
+      ssl_mutexes[mutex_num].lock_shared();
+    } else if (mode & CRYPTO_UNLOCK) {
+      ssl_mutexes[mutex_num].unlock_shared();
+    }
+  } else if (mode & CRYPTO_WRITE) {
+    if (mode & CRYPTO_LOCK) {
+      ssl_mutexes[mutex_num].lock();
+    } else if (mode & CRYPTO_UNLOCK) {
+      ssl_mutexes[mutex_num].unlock();
     }
-    crypto_context = NSS_InitContext(cct->_conf->nss_db_path.c_str(), "", "",
-                                     SECMOD_DB, &init_params, flags);
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
-  ceph_assert_always(crypto_context != NULL);
 }
 
-void ceph::crypto::shutdown(bool shared)
+static unsigned long
+ssl_get_thread_id(void)
 {
-  pthread_mutex_lock(&crypto_init_mutex);
-  ceph_assert_always(crypto_refs > 0);
-  if (--crypto_refs == 0) {
-    NSS_ShutdownContext(crypto_context);
-    if (!shared) {
-      PR_Cleanup();
-    }
-    crypto_context = NULL;
-    crypto_init_pid = 0;
+  static_assert(sizeof(unsigned long) >= sizeof(pthread_t));
+  /* pthread_t may be any data type, so a simple cast to unsigned long
+   * can rise a warning/error, depending on the platform.
+   * Here memcpy is used as an anything-to-anything cast. */
+  unsigned long ret = 0;
+  pthread_t t = pthread_self();
+  memcpy(&ret, &t, sizeof(pthread_t));
+  return ret;
+}
+#endif /* not OPENSSL_API_1_1 */
+
+static void init() {
+#ifndef OPENSSL_API_1_1
+  if (++crypto_refs != 1) {
+    return;
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
+
+  /* According to https://wiki.openssl.org/index.php/Library_Initialization#libcrypto_Initialization */
+  OpenSSL_add_all_algorithms();
+  ERR_load_crypto_strings();
+
+  /* Initialize locking callbacks, needed for thread safety.
+   * http://www.openssl.org/support/faq.html#PROG1
+   */
+  CRYPTO_set_locking_callback(&ssl_locking_callback);
+  CRYPTO_set_id_callback(&ssl_get_thread_id);
+
+  OPENSSL_config(nullptr);
```

Comment:
per https://www.openssl.org/docs/man1.0.2/man3/OPENSSL_config.html, we need to call `CONF_modules_free()` to free up the memory allocated by all configuration modules.  following valgrind failures echo this suggestion:

- http://pulpito.ceph.com/kchai-2019-06-20_05:33:07-rados-wip-kefu-testing-2019-06-20-1103-distro-basic-smithi/4050060/
- http://pulpito.ceph.com/kchai-2019-06-20_05:33:07-rados-wip-kefu-testing-2019-06-20-1103-distro-basic-smithi/4050175/
- http://pulpito.ceph.com/kchai-2019-06-20_05:33:07-rados-wip-kefu-testing-2019-06-20-1103-distro-basic-smithi/4050291/

```xml
  <stack>
    <frame>
      <ip>0x4C29BC3</ip>
      <obj>/usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>malloc</fn>
      <dir>/builddir/build/BUILD/valgrind-3.13.0/coregrind/m_replacemalloc</dir>
      <file>vg_replace_malloc.c</file>
      <line>299</line>
    </frame>
    <frame>
      <ip>0xF3B54B7</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
      <fn>CRYPTO_malloc</fn>
    </frame>
    <frame>
      <ip>0xF460D01</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
      <fn>BUF_strndup</fn>
    </frame>
    <frame>
      <ip>0xF4C5F1A</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
    </frame>
    <frame>
      <ip>0xF4C65E3</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
      <fn>CONF_module_add</fn>
    </frame>
    <frame>
      <ip>0xF4C696D</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
      <fn>OPENSSL_load_builtin_modules</fn>
    </frame>
    <frame>
      <ip>0xF4C6998</ip>
      <obj>/usr/lib64/libcrypto.so.1.0.2k</obj>
      <fn>OPENSSL_config</fn>
    </frame>
    <frame>
      <ip>0x51E0F22</ip>
      <obj>/usr/lib64/ceph/libceph-common.so.0</obj>
      <fn>common_init_finish(CephContext*)</fn>
      <dir>/usr/src/debug/ceph-15.0.0-2031-g5348b4c/src/common</dir>
      <file>common_init.cc</file>
      <line>94</line>
    </frame>
```